framework/sandstone.cpp

```diff
@@ -1773,6 +1773,44 @@ static void run_one_test_init_in_parent(ChildrenList &children, const struct tes
     children.results.emplace_back(ChildExitStatus{ res });
 }
 
+static void run_one_test_in_parent(ChildrenList &children, const struct test *test)
+{
+    TestResult state = TestResult::Passed;
+    init_per_thread_data();
+
+    do {
+        if (test->test_init) {
+            int ret = test->test_init(const_cast<struct test *>(test));
+            PerThreadData::Main *thr = sApp->main_thread_data();
+            if (ret > EXIT_SUCCESS || thr->has_failed()) {
+                state = TestResult::Failed;
+                break;
+            } else if (ret < EXIT_SUCCESS || thr->has_skipped()) {
+                state = TestResult::Skipped;
+                break;
+            }
+        }
+
+        run_threads(test);
+
+        if (test->test_cleanup) {
+            int ret = test->test_cleanup(const_cast<struct test *>(test));
+            PerThreadData::Main *thr = sApp->main_thread_data();
+            if (ret > EXIT_SUCCESS || thr->has_failed()) {
+                state = TestResult::Failed;
+            } else if (ret == EXIT_SKIP || thr->has_skipped()) {
+                log_skip(RuntimeSkipCategory, "SKIP requested in cleanup");
+                state = TestResult::Skipped;
+            } else if (ret < EXIT_SUCCESS) {
+                log_skip(RuntimeSkipCategory, "Unexpected OS error in cleanup: %s", strerror(-ret));
+                state = TestResult::Skipped;
+            }
+        }
+    } while(false);
```

Comment:
This is the main portion of `child_run()`, with the difference that the init function is run in the main process' main thread, instead of an auxiliary thread. Instead of C&P, please merge the two, in a function named something like `inner_run_one_test` or `run_one_test_inner`.